src/VisualStudio/Core/Def/RoslynPackage.cs

```diff
@@ -93,8 +95,15 @@ private void InitializeColors()
             CodeAnalysisColors.AccentBarColorKey = EnvironmentColors.FileTabInactiveDocumentBorderEdgeBrushKey;
         }
 
-        protected override void LoadComponentsInUIContext(CancellationToken cancellationToken)
+        protected override async Task LoadComponentsInUIContextAsync(CancellationToken cancellationToken)
         {
+            await GetServiceAsync(typeof(SVsTaskStatusCenterService));
```

Comment:
AbstractPackage2 switched to the UI thread as teh first thing.  But this code did not.  intentional?

Reply:
I'm looking for some good, clear, rules about the exact patterns we should have for async.  Specifically for JTFish code (since i think i have a good grasp on the roslyn rules for non-jtf code).  It's hard to review this stuff since i can't tell what is intentional and part of an accepted pattern, vs waht may be a mistake by omission.

Reply:
Can we tweak this at all:

> Async UI code and the package initialization code should use SwitchToMainThreadAsync (if necessary) and omit ConfigureAwait overall

Could we make that:

> Async UI code and the package initialization code should use `SwitchToMainThreadAsync` (if necessary) and use `.ConfigureAwait(true)` overall

Having the ConfigureAwait be explicit in teh code makes it clear that the behavior is *intentional* and not he result of accidental omission.  It's the same reason we ask that CancellationToken.None be explicitly passed instead of having that happen because of a default-parameter-value.  By being explicit, someone looking at the code (1, 2, 5, years down the line) can reasonable assert: yes, the intent is clear here.  there wasn't an accidental omission.

This sort of explicitness has been *extremely* valuable for us in the past.  It has helped catch a bunch of issues, and it makes reviewing much simpler since it takes the guesswork out of things.  It does make things more verbose.  But i think that's a virtue here as being overly clear is better for these subtle areas.

Reply:
Each asynchronous method that requires code to execute on the main thread should use `SwitchToMainThreadAsync`. It *should not* assume that the caller has done so. This results in explicit local statement of affinity and eliminates the need to use `ConfigureAwait(true)`.

Code that does not require the UI thread would fall into the category of using `ConfigureAwait(false)`.

Reply:
I see vs-threading as the writing on the wall for the eventually elimination of widespread littering of `ConfigureAwait`. We aren't ready to go down that road yet, but it's the target outcome.

Reply:
> Each asynchronous method that requires code to execute on the main thread should use SwitchToMainThreadAsync

But this method doesn't do that... 

And it has code that says this:

```
             // The misc files workspace needs to be loaded on the UI thread.  This way it will have
             // the appropriate task scheduler to report events on.
             this.ComponentModel.GetService<MiscellaneousFilesWorkspace>();
```

So i'm still stuck trying to figure out if this is correct or not... :-/

Reply:
> So i'm still stuck trying to figure out if this is correct or not... :-/

It's not, it should have an explicit switch before that line.

Reply:
> I see vs-threading as the writing on the wall for the eventually elimination of widespread littering of ConfigureAwait. We aren't ready to go down that road yet, but it's the target outcome.

I don't like preemptively changing styles when it's not been guaranteed what the final state will be.  In this interim period, for example, it does lead to confusion (i.e. both these threads where i'm still not sure if this code is correct :)).  

I'm not trying to hold anything back.  I'm literally just trying to figure out what the rules are and how to reason about this code given those rules.  Right now it's not at all clear to me (even if i take a stance of: "AbstractPackage is special and goes its own way").  

Let's just look at this method on its own.  First, i have no idea how to reason about what thread it is actually on. It says LoadComponentsInUIContextAsync.  However, there is no assert to actualy verify that it's on the UI context.  Then, it performs a bunch of async calls that are themselves not clear to me as to their intent (though i can come aronud around to the idea that if you knwo this is JTF, you can maybe squint and view no-configure-await as being intentional).  Then, it runs code that affectively says "if i'm not on the UI thread, bad baddy badness will occur", but i see nothing that ensures that safety.

Overall, i find this pattern highly confusing and very difficult to reason about.  Invariants are not made clear, and it seems far too easy to accidentally do the wrong thing.

Reply:
First of all, the discussion is great. It helps highlight things that aren't obvious.

>  It says LoadComponentsInUIContextAsync.

I renamed the method to `LoadComponentsAsync`. It's weird because `UIContext` here is a concept independent from the concept of "main thread". Asynchronous methods shouldn't use the method name as an indicator of thread affinity, so we should come up with a new name for this method to avoid confusion.

> Then, it performs a bunch of async calls that are themselves not clear to me as to their intent

I am very much interested in your opinion of whether the direction of sharwell/roslyn@ded603f29bbdd8f67aed6920cce9e0e527ddc529 makes things clearer (not considering its current lack of documentation).

> Then, it runs code that affectively says "if i'm not on the UI thread, bad baddy badness will occur", but i see nothing that ensures that safety.

This is breaking the rules. Affinity should be explicit, not implicit.

> ... difficult to reason about ... Invariants are not made clear ...

If we add in the explicit `SwitchToMainThreadAsync`, does it address the confusion within this context?

Reply:
> I am very much interested in your opinion of whether the direction of sharwell/roslyn@ded603f makes things clearer (not considering its current lack of documentation).

Overall, yes.  I prefer declarative dependency declarations.  Though i did have some comments on that PR around clarity.

Reply:
@jinujoseph @dotnet/roslyn-ide I think whether we want .ConfigureAwait(true) to be always added or not is debatable. probably better to be discussed in team meeting.